simpleclient/src/main/java/io/prometheus/client/Collector.java

```diff
@@ -19,27 +20,58 @@
    */
   public abstract List<MetricFamilySamples> collect();
   public enum Type {
+    UNKNOWN, // This is untyped in Prometheus text format.
     COUNTER,
     GAUGE,
-    SUMMARY,
+    STATE_SET,
+    INFO,
     HISTOGRAM,
-    UNTYPED,
+    GAUGE_HISTOGRAM,
+    SUMMARY,
   }
 
   /**
    * A metric, and all of its samples.
    */
   static public class MetricFamilySamples {
     public final String name;
+    public final String unit;
     public final Type type;
     public final String help;
     public final List<Sample> samples;
 
-    public MetricFamilySamples(String name, Type type, String help, List<Sample> samples) {
+    public MetricFamilySamples(String name, String unit, Type type, String help, List<Sample> samples) {
+      if (!unit.isEmpty() && !name.endsWith("_" + unit)) {
+        throw new IllegalArgumentException("Metric's unit is not the suffix of the metric name: " + name);
+      }
+      if ((type == Type.INFO || type == Type.STATE_SET) && !unit.isEmpty()) {
+        throw new IllegalArgumentException("Metric is of a type that cannot have a unit: " + name);
+      }
+      List<Sample> mungedSamples = samples;
+      // Deal with _total from pre-OM automatically.
+      if (type == Type.COUNTER) {
+        if (name.endsWith("_total")) {
```

Comment:
nit: Worth making `_total` a const? Then can `name.endsWith(counterSuffix)` and `name = name.substring(0, name.length() - counterSuffix.length())`.

Reply:
I don't like indirection like that, it only makes the code harder to understand for me as I have to bounce around the file for something that could be understood one its own line.